cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go

```diff
@@ -26,73 +26,179 @@ import (
+	// fields in FieldSelector is very small. things like "spec.chart" or "status.artifact.revision" are
+	// certainly not supported.
+	// see
+	//  - kubernetes/client-go#713 and
+	//  - https://github.com/flant/shell-operator/blob/8fa3c3b8cfeb1ddb37b070b7a871561fdffe788b///HOOKS.md#fieldselector and
+	//  - https://github.com/kubernetes/kubernetes/issues/53459
+	chartList, err := s.listChartsInCluster(ctx, namespace)
+	if err != nil {
+		return "", err, nil
 	}
 
-	conditions, found, err := unstructured.NestedSlice(unstructuredChart.Object, "status", "conditions")
+	url, err = findUrlForChartInList(chartList, repoName, chartName, chartVersion)
 	if err != nil {
-		return false, err
-	} else if !found {
-		return false, nil
-	}
-
-	// check if ready=True
-	for _, conditionUnstructured := range conditions {
-		if conditionAsMap, ok := conditionUnstructured.(map[string]interface{}); ok {
-			if typeString, ok := conditionAsMap["type"]; ok && typeString == "Ready" {
-				if statusString, ok := conditionAsMap["status"]; ok {
-					if statusString == "True" {
-						if reasonString, ok := conditionAsMap["reason"]; !ok || reasonString != "ChartPullSucceeded" {
-							// should not happen
-							log.Infof("unexpected status of HelmChart: %v", *unstructuredChart)
-						}
-						return true, nil
-					} else if statusString == "False" {
-						var msg string
-						if msg, ok = conditionAsMap["message"].(string); !ok {
-							msg = fmt.Sprintf("No message available in condition: %v", conditionAsMap)
-						}
-						// chart pull is done and it's a failure
-						return true, status.Errorf(codes.Internal, msg)
-					}
+		return "", err, nil
+	} else if url != "" {
+		return url, nil, nil
+	}
+
+	// did not find the chart, need to create
+	// see https://fluxcd.io/docs/components/source/helmcharts/
+	// TODO (gfichtenholt)
+	// 1. HelmChart object needs to be co-located in the same namespace as the HelmRepository it is referencing.
+	// 2. flux impersonates a "super" user when doing this (see fluxv2 plug-in specific notes at the end of
+	//	design doc). We should probably be doing simething similar to avoid RBAC-related problems
+	unstructuredChart := newFluxHelmChart(chartName, repoName, chartVersion)
+
+	resourceIfc, err := s.getChartsResourceInterface(ctx, namespace)
+	if err != nil {
+		return "", err, nil
+	}
+
+	newChart, err := resourceIfc.Create(ctx, &unstructuredChart, metav1.CreateOptions{})
+	if err != nil {
+		log.Errorf("Error creating chart: %v\n%v", err, unstructuredChart)
+		return "", err, nil
+	}
+
+	log.Infof("Created chart: [%v]", prettyPrintMap(newChart.Object))
+
+	// Delete the created helm chart regardless of success or failure. At the end of
+	// GetAvailablePackageDetail(), we've already collected the information we need,
+	// so why leave a flux chart chart object hanging around?
+	// Over time, they could accumulate to a very large number...
```

Comment:
Yes. More generally it seems a bit icky to be creating them when they should exist if a HelmRelease already exists (and if not, I'm not sure we should be trying to provide the info). It seems analogous to the situation with normal helm, where if the user has deleted the AppRepository (ie. so kubeapps no longer has info about the related charts), we can't provide that info (such as upgrade info or latest version) and that's OK.

Reply:
GetAvailablePackageDetail() requires me to provide some info not available otherwise, like the readme. For a chart that may not be installed at all. I can only get that if I have the chart tgz file, so that's the only reason I am doing it. I am open to other proposals, please.

Reply:
Yep, it makes sense for GetAvailablePackageDetail() - and given that's a singular request, the efficiency isn't an issue either. As above, I'd thought this PR was about GetInstalledPackageSummaries only as per the title, so when I saw these changes, assumed it must be for them.